src/aspire/source/image.py

```diff
@@ -597,16 +596,68 @@ def save(
             overwrite=overwrite,
         )
 
-    def save_metadata(
-        self, starfile_filepath, new_mrcs=True, batch_size=512, save_mode=None
+    def _populate_common_metadata(
+        self,
+        df,
+        local_cols,
+        starfile_filepath,
+        batch_size,
+        save_mode,
     ):
+        """
+        Populate metadata columns common to all `ImageSource` subclasses.
+        """
+        # Create a new column that we will be populating in the loop below
+        df["_rlnImageName"] = ""
+
+        if save_mode == "single":
+            # Save all images into one single mrc file
+            fname = os.path.basename(starfile_filepath)
+            fstem = os.path.splitext(fname)[0]
+            mrcs_filename = f"{fstem}_{0}_{self.n-1}.mrcs"
+
+            # Then set name in dataframe for the StarFile
+            # Note, here the row_indexer is :, representing all rows in this data frame.
+            #   df.loc will be reponsible for dereferencing and assigning values to df.
+            #   Pandas will assert df.shape[0] == self.n
+            df.loc[:, "_rlnImageName"] = [
+                f"{j + 1:06}@{mrcs_filename}" for j in range(self.n)
+            ]
+        else:
+            # save all images into multiple mrc files in batch size
+            for i_start in np.arange(0, self.n, batch_size):
+                i_end = min(self.n, i_start + batch_size)
+                num = i_end - i_start
+                mrcs_filename = (
+                    os.path.splitext(os.path.basename(starfile_filepath))[0]
+                    + f"_{i_start}_{i_end-1}.mrcs"
+                )
+                # Note, here the row_indexer is a slice.
+                #   df.loc will be reponsible for dereferencing and assigning values to df.
+                #   Pandas will assert the lnegth of row_indexer equals num.
+                row_indexer = df[i_start:i_end].index
+                df.loc[row_indexer, "_rlnImageName"] = [
+                    "{0:06}@{1}".format(j + 1, mrcs_filename) for j in range(num)
+                ]
+
+        # Subclass-specific columns are popped to the end of the dataframe
+        for col in local_cols:
+            df[col] = df.pop(col)
```

Comment:
Double checking I understand what is happening here.

The RHS returns the column from df, and removes it from df.
The LHS assigns that as a new column.

Net effect is that the columns are reordered.

Reply:
Exactly, `pop()` returns a Series and drops it from the column.

Reply:
Looks good, thanks